core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java

```diff
@@ -153,40 +129,101 @@ public final void success(CoreEvent event) {
             .then()
             .toProcessor();
       } else {
-        responseProcessor.onError(throwable);
+        responseDone(null, throwable);
         return empty();
       }
     }
   }
 
+  private synchronized void responseDone(CoreEvent event, Throwable throwable) {
+    this.state = STATE_DONE;
+    onResponseConsumerList.stream().forEach(consumer -> consumer.accept(event, throwable));
+    tryComplete();
+    if (throwable != null) {
+      responseProcessor.onError(throwable);
+    } else {
+      responseProcessor.onNext(event);
+    }
+    tryTerminate();
+  }
+
+  protected synchronized void tryComplete() {
+    if (this.state == STATE_DONE && childContexts.stream().filter(context -> !context.isComplete()).count() == 0) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(this + " completed.");
+      }
+      this.state = STATE_COMPLETE;
+      if (responseProcessor.isError()) {
+        onCompletionConsumerList.forEach(runnable -> runnable.accept(null, responseProcessor.getError()));
+      } else {
+        onCompletionConsumerList.forEach(consumer -> consumer.accept(responseProcessor.peek(), null));
+      }
+      getParentContext().ifPresent(context -> {
+        if (context instanceof AbstractEventContext) {
+          ((AbstractEventContext) context).tryComplete();
+        }
+      });
+      tryTerminate();
+    }
+  }
+
+  protected synchronized void tryTerminate() {
+    if (this.state == STATE_COMPLETE && externalCompletion) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(this + " terminated.");
+      }
+      this.state = STATE_TERMINATED;
+      if (responseProcessor.isError()) {
+        onTerminatedConsumerList.forEach(runnable -> runnable.accept(null, responseProcessor.getError()));
+      } else {
+        onTerminatedConsumerList.forEach(consumer -> consumer.accept(responseProcessor.peek(), null));
+      }
+    }
+  }
+
   @Override
   public BaseEventContext getRootContext() {
     return getParentContext()
         .map(BaseEventContext::getRootContext)
         .orElse(this);
   }
 
+  protected FlowExceptionHandler getExceptionHandler() {
+    return exceptionHandler;
+  }
+
   @Override
-  public Publisher<CoreEvent> getBeforeResponsePublisher() {
-    return beforeResponseProcessor;
+  public boolean isComplete() {
+    return this.state >= STATE_COMPLETE;
   }
 
   @Override
-  public Publisher<CoreEvent> getResponsePublisher() {
-    return responseProcessor;
+  public boolean isTerminated() {
+    return this.state == STATE_TERMINATED;
   }
 
   @Override
-  public Publisher<Void> getCompletionPublisher() {
-    return completionProcessor;
+  public synchronized void onTerminated(BiConsumer<CoreEvent, Throwable> consumer) {
+    onTerminatedConsumerList.add(consumer);
   }
 
-  protected FlowExceptionHandler getExceptionHandler() {
-    return exceptionHandler;
+  @Override
+  public synchronized void onComplete(BiConsumer<CoreEvent, Throwable> consumer) {
```

Comment:
why do you add this but didn't change all the uses of `from(eventContext.getCompletionPublisher())`? For example, look at `org.mule.runtime.core.internal.streaming.CursorManager#registerEventContext`, shouldn't it be using this new method as well?... In either way, seems to me like now I have two ways of doing the same thing

Reply:
Did you not see these changes:
https://github.com/mulesoft/mule/pull/6147/files#diff-8ab0c830035b41ab12017e91072885bdR137
https://github.com/mulesoft/mule/pull/6147/files#diff-c7ba2cf2082917d74cf6465bc46d16bbL185

Reply:
i clearly did not 😱